proposals/NNNN-ComparisonReform.md

```diff
@@ -0,0 +1,531 @@
+* Status: **Awaiting review**
+* Review manager: TBD
+
+## Introduction
+
+This proposal is for changes that we believe should be made to the existing
+comparison system by:
+
+* Making FloatingPoint comparison context sensitive, so that its `Comparable`
+  conformance provides a proper total ordering. 
+* Introducing a new ternary-valued `compare(_ other: Self) -> ComparisonResult`
+  method. 
+* Removing unnecessary customization points from `Comparable`.
+
+## Motivation
+
+The motivation comes from several independent points:
+
+1: The standard comparison operators have an intuitive meaning to programmers.
+Swift encourages encoding that in an implementation of `Comparable` that
+respects the rules of a [total
+order](https://en.wikipedia.org/wiki/Total_order). The standard library takes
+advantage of these rules to provide consistent implementations for sorting and
+searching generic collections of `Comparable` types.
+
+Not all types behave so well in this framework, unfortunately. There are cases
+where the semantics of a total order cannot apply while still maintaining the
+traditional definition of “comparison” for these types. Take, for example,
+sorting an array of `Float`s. Today, `Float`'s instance of `Comparable` follows
+IEEE-754 and returns `false` for all comparisons of `NaN`. In order to sort
+this array, `NaN` s are considered outside the domain of `<`, and the order of
+a sorted array containing them is unspecified. Similarly, a Dictionary keyed
+off floats can leak entries and memory.
+
+2: Generic algorithms in the Swift Standard Library that make use of the
+current `Comparable` protocol may have to make extra comparisons to determine
+the ordering of values when `<`, `==`, and `>` should have different
+behaviours. Having a central operation to return complete ordering information
+should provide a speedup for these operations.
+
+3: The existing comparison operators don't "generalize" well. There's no clean
+way to add a third or fourth argument to `<` to ask for non-default semantics.
+An example where this would be desirable would be specifying the locale or
+case-sensitivity when comparing Strings.
+
+4: `Comparable` is over-engineered in the customization points it provides: to
+our knowledge, there's no good reason to ever override `>=`, `>`, or `<=`. Each
+customization point bloats vtables and mandates additional dynamic dispatch.
+
+5: When quickly writing a `Comparable` type, it is easier to implement a single
+ternary statement than to separately implement `==` and `<`.
+
+## Proposed solution
+
+### `ComparisonResult`
+
+Foundation's ComparisonResult type will be mapped into Swift as 
+
+```swift
+@objc public enum ComparisonResult : Int {
+  case orderedAscending = -1
+  case orderedSame = 0
+  case orderedDescending = 1
+}
+```
+
+
+### `Comparable`
+
+Comparable will be changed to have a new ternary comparison method: `compare(_
+other: Self) -> ComparisonResult`. `x.compare(y)` specifies where to place x
```

Comment:
What is the motivation for an instance method over a static method? I realize that `NSObject` sets a precedent here, but I'd like it to be explained before we copy that design.

I personally feel that both the left-hand side and the right-hand side of a comparison should be on equal grounding. An instance method privileges the left-hand argument with dynamic dispatch. Maybe this is desirable behavior for classes? I'm not convinced.

I would prefer the following design of `static`/`class` function as well as an additional top-level function:
```swift
protocol Comparable {
  static func compare(_ lhs: Self, _ rhs: Self) -> ComparisonResult
}

fun compare<T: Comparable>(_ lhs: T, _ rhs: T) -> ComparisonResult {
  return T.compare(lhs, rhs)
}
```

This design removes the asymmetry since `compare` is no longer an instance method.
```swift
switch compare(a, b) { ... }
```

Notably, this design results in static dispatch by default when comparing two values since `T` will be unified to the closest ancestor class of the arguments. If dynamic dispatch is desired (potentially for certain OOP designs), the implementor can delegate to an appropriate subclassable method within `compare`.